business/jaeger_helper.go

```diff
@@ -84,6 +100,143 @@ func getErrorTracesFromJaeger(namespace string, service string, requestToken str
+}
+
+func queryTraces(u *url.URL, auth config.Auth, jaegerQ url.Values) ([]jaeger.Trace, error) {
+	u.RawQuery = jaegerQ.Encode()
+	log.Infof("Jaeger query: %s", u)
+	body, code, reqError := httputil.HttpGet(u.String(), &auth, 5*time.Second)
+	if reqError != nil {
+		log.Errorf("Error fetching Jaeger Error Traces (%d): %s", code, reqError)
+		return []jaeger.Trace{}, reqError
+	}
+
+	if code != http.StatusOK {
+		return []jaeger.Trace{}, fmt.Errorf("error from Jaeger (%d)", code)
+	}
+
+	var traces RequestTrace
+	if errMarshal := json.Unmarshal(body, &traces); errMarshal != nil {
+		log.Errorf("Error unmarshalling Jaeger response: %s", errMarshal)
+		return []jaeger.Trace{}, errMarshal
+	}
+
+	return traces.Traces, nil
+}
+
+func tracesToSpans(traces []jaeger.Trace, service, namespace string) []Span {
+	spans := []Span{}
+	for _, trace := range traces {
+		// First, get the desired processes for our service
+		processes := make(map[jaeger.ProcessID]bool)
+		for pId, process := range trace.Processes {
+			if process.ServiceName == service || process.ServiceName == service+"."+namespace {
+				processes[pId] = true
+			}
+		}
+		// Second, find spans for these processes
+		for _, span := range trace.Spans {
+			if ok := processes[span.ProcessID]; ok {
+				spans = append(spans, Span{
+					Span:      span,
+					TraceSize: len(trace.Spans),
+				})
+			}
+		}
+	}
+	log.Infof("Found %d spans in the %d traces for service %s", len(spans), len(traces), service)
+	return spans
+}
+
+func findRelevantSpans(spansSample []Span, u *url.URL, auth config.Auth, jaegerQ url.Values, service, namespace string) ([]Span, error) {
+	spansMap := make(map[jaeger.SpanID]Span)
+
+	// Query for errors
+	jaegerQ.Set("tags", "{\"error\":\"true\"}")
+	traces, _ := queryTraces(u, auth, jaegerQ)
+	errSpans := tracesToSpans(traces, service, namespace)
+	for _, span := range errSpans {
+		spansMap[span.SpanID] = span
+	}
+
+	// Find 90th percentile; sort per duration
+	sort.Slice(spansSample, func(i, j int) bool {
+		return spansSample[i].Span.Duration < spansSample[j].Span.Duration
+	})
+	idx90 := int(9 * len(spansSample) / 10)
+	duration90th := time.Duration(spansSample[idx90].Duration) * time.Microsecond
+	log.Infof("90th percentile duration: %s", duration90th)
+	for _, span := range spansSample[idx90:] {
+		spansMap[span.SpanID] = span
+	}
+
+	// Query 90th percentile
+	jaegerQ.Del("tags")
+	// %.1gms would print for instance 0.00012456 as 0.0001ms
+	jaegerQ.Set("minDuration", fmt.Sprintf("%.1gms", float64(duration90th.Nanoseconds())/1000000))
+	traces, _ = queryTraces(u, auth, jaegerQ)
+	// TODO / Question: if limit is reached again we might limit to 99th percentile instead?
+	pct90Spans := tracesToSpans(traces, service, namespace)
+	for _, span := range pct90Spans {
+		spansMap[span.SpanID] = span
+	}
+
+	// Map to list
+	ret := []Span{}
+	for _, span := range spansMap {
+		ret = append(ret, span)
+	}
+	log.Infof("Found %d relevant spans", len(ret))
+	return ret, nil
+}
```

Comment:
@jpkrohling @objectiser when the limit of traces has been reached (limit set to 100 at the moment) I'm trying here to fetch more relevant spans, taking the ones in error + the ones with duration above 90th percentile. WDYT? Does the algorithm sound ok ? Or do you have better suggestion?

Reply:
I think I'm missing some context. Are you then retrieving the traces where the relevant spans occurred? What makes a span more relevant?

Reply:
@jpkrohling Well, I might be wrong, but on the UI we will display Spans and not Traces (see https://github.com/kiali/kiali-ui/pull/1560 ) - only the Span has really a duration, and is directly correlated to a Service, isn't it?

So, my endpoint here will be returning a list of spans to be displayed in the UI. Note that I'm adding the trace size (number of spans) within my span model so that it display smaller or bigger circles accordingly, hence keeping the notion of trace size. The intent is also to be able to jump to a Trace from a Span, maybe with a link to Jaeger UI, although it's not for this PR / will come later.

Reply:
I think I see what you mean now. One thing to consider: should you be selecting only the proxy spans?

Reply:
@jpkrohling I don't think we should select only the proxy spans, as correlating user-defined spans also brings value I think.
However we still perform some selection as we're using the service name as defined by the proxies, which formats as "service.namespace". So if user-defined traces or spans use a different format, they wouldn't be caught here. We can probably enhance this later.